mixnet/queuesim/simulation.py

```diff
@@ -0,0 +1,142 @@
+        # To collect the dissemination times of each message.
+        dissemination_times: list[float] = []
+        # Wait until all messages are disseminated to the entire network.
+        while len(dissemination_times) < self.config.num_sent_msgs:
+            # Wait until a node notifies that it has received a new message.
+            received_time, msg = await self.received_msg_queue.get()
+            # If the message has been received by all nodes, calculate the dissemination time.
+            received_msg_counters.update([msg])
+            if received_msg_counters[msg] == len(nodes):
+                dissemination_times.append(
+                    received_time - Message.from_bytes(msg).sent_time
+                )
+        return dissemination_times
+
+    def __run_nodes(self) -> list[Node]:
+        return [
+            Node(
+                self.framework,
+                self.config.nomssip,
+                self.__process_msg,
+            )
+            for _ in range(self.config.num_nodes)
+        ]
+
+    async def __process_msg(self, msg: bytes) -> None:
+        """
+        A handler to process messages received via Nomos Gossip channel
+        """
+        # Notify that a new message has been received by the node.
+        # The received time is also included in the notification.
+        await self.received_msg_queue.put((self.framework.now(), msg))
+
+    def __connect_nodes(self, nodes: list[Node]):
+        topology = build_full_random_topology(
+            rng=self.config.topology.seed,
+            num_nodes=len(nodes),
+            peering_degree=self.config.nomssip.peering_degree,
+        )
+        meter_start_time = self.framework.now()
+        # Sort the topology by node index for the connection RULE defined below.
+        for node_idx, peer_indices in sorted(topology.items()):
+            for peer_idx in peer_indices:
+                # Since the topology is undirected, we only need to connect the two nodes once.
+                # RULE: the node with the smaller index establishes the connection.
+                assert node_idx != peer_idx
+                if node_idx > peer_idx:
+                    continue
+
+                # Connect the node and peer for Nomos Gossip
+                node = nodes[node_idx]
+                peer = nodes[peer_idx]
+                node.connect(
+                    peer,
+                    self.__create_conn(meter_start_time),
+                    self.__create_conn(meter_start_time),
+                )
+
+    def __create_conn(self, meter_start_time: float) -> MeteredRemoteSimplexConnection:
+        return MeteredRemoteSimplexConnection(
+            self.config.latency,
+            self.framework,
+            meter_start_time,
+        )
+
+    async def __run_sender(self, sender: Node):
+        for i in range(self.config.num_sent_msgs):
+            if i > 0:
+                await self.framework.sleep(self.config.msg_interval_sec)
```

Comment:
Do we really wait `sleep` here? Do we `await` for `msg_interval_sec` of time?

If yes then that is going to be a major blocker in terms of execution speed of experiments.

Therefore, I would suggest (if that is feasible) to simulate the time. We can work with discreet timing of abstract `unit_of_time` (which can be a second or any other time unit). As we are interested in measuring message dissemination time we can assume that our clock is following the transmission rate. That is, each transmission takes a single `unit_of_time`.

We are not interested in the processing delay, so we can safely assume it is 0. Later we can set it to something larger, which we can express in the same `unit_of_time`.

Reply:
We are already using the [discrete virtual time](https://usim.readthedocs.io/en/latest/source/api/usim.html#simulating-time) implemented by [μSim](https://usim.readthedocs.io/en/latest/index.html). So, even if we call `await self.framework.sleep(1)`, it doesn't wait for 1 second in real time. How long the single time unit `1` will take depends on how many tasks are scheduled around the specific virtual time, and how fast the CPU is. 

Nevertheless, I'm using the name `_sec` because we're calculate bandwidth usages in `KiB/s`. So, I'm considering a single virtual time `1` as 1 second in real time, just as a notation, because someone might be curious what `/s` means in the world of discrete virtual time. If you think that it's confusing, I can remove the `_sec` suffix. 

> We are not interested in the processing delay, so we can safely assume it is 0. Later we can set it to something larger, which we can express in the same unit_of_time.

Yes. Any processing delay (such as computation time) is not included in the calculation of dissemination time. Only queuing time and network latency are included in the dissemination time, thanks to the discrete virtual time. 
(In this experiment, I set network latency to 0).

Reply:
Great! That's ok, no need to remove the `_sec`. I was a bit confused seeing that `sleep` and I have overlooked the fact that it comes from the `usim`. Thanks for clarification!